package dev.eness.sololevelingfinal.core.client.model;

import com.mojang.blaze3d.vertex.PoseStack;
import com.mojang.blaze3d.vertex.VertexConsumer;
import net.minecraft.client.model.EntityModel;
import net.minecraft.client.model.geom.ModelLayerLocation;
import net.minecraft.client.model.geom.ModelPart;
import net.minecraft.client.model.geom.PartPose;
import net.minecraft.client.model.geom.builders.CubeDeformation;
import net.minecraft.client.model.geom.builders.CubeListBuilder;
import net.minecraft.client.model.geom.builders.LayerDefinition;
import net.minecraft.client.model.geom.builders.MeshDefinition;
import net.minecraft.client.model.geom.builders.PartDefinition;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.world.entity.Entity;

public class Modelmanaarrow<T extends Entity> extends EntityModel<T> {
   public static final ModelLayerLocation LAYER_LOCATION = new ModelLayerLocation(new ResourceLocation("sololeveling", "modelmanaarrow"), "main");
   public final ModelPart main;

   public Modelmanaarrow(ModelPart root) {
      this.main = root.getChild("main");
   }

   public static LayerDefinition createBodyLayer() {
      MeshDefinition meshdefinition = new MeshDefinition();
      PartDefinition partdefinition = meshdefinition.getRoot();
      PartDefinition main = partdefinition.addOrReplaceChild(
         "main",
         CubeListBuilder.create()
            .texOffs(0, 1)
            .addBox(0.0F, -0.5F, 0.0F, 7.0F, 1.0F, 0.0F, new CubeDeformation(0.0F))
            .texOffs(4, 4)
            .addBox(-2.0F, -0.5F, -0.5F, 2.0F, 0.0F, 1.0F, new CubeDeformation(0.0F))
            .texOffs(0, 4)
            .addBox(-2.0F, 0.5F, -0.5F, 2.0F, 0.0F, 1.0F, new CubeDeformation(0.0F))
            .texOffs(4, 5)
            .addBox(-2.0F, -0.5F, 0.5F, 2.0F, 1.0F, 0.0F, new CubeDeformation(0.0F))
            .texOffs(0, 5)
            .addBox(-2.0F, -0.5F, -0.5F, 2.0F, 1.0F, 0.0F, new CubeDeformation(0.0F))
            .texOffs(-1, 0)
            .addBox(0.0F, 0.0F, -0.5F, 7.0F, 0.0F, 1.0F, new CubeDeformation(0.0F)),
         PartPose.offsetAndRotation(0.75F, 18.5F, 0.0F, 0.0F, 0.0F, -1.5708F)
      );
      PartDefinition cube_r1 = main.addOrReplaceChild(
         "cube_r1",
         CubeListBuilder.create().texOffs(0, 2).addBox(-1.0F, 0.0F, -0.5F, 3.0F, 0.0F, 1.0F, new CubeDeformation(0.0F)),
         PartPose.offsetAndRotation(-2.0F, 0.0F, 0.0F, -2.3562F, 0.0F, 0.0F)
      );
      PartDefinition cube_r2 = main.addOrReplaceChild(
         "cube_r2",
         CubeListBuilder.create().texOffs(0, 3).addBox(-1.0F, 0.0F, -0.5F, 3.0F, 0.0F, 1.0F, new CubeDeformation(0.0F)),
         PartPose.offsetAndRotation(-2.0F, 0.0F, 0.0F, -0.7854F, 0.0F, 0.0F)
      );
      return LayerDefinition.create(meshdefinition, 16, 16);
   }

   @Override
   public void setupAnim(T entity, float limbSwing, float limbSwingAmount, float ageInTicks, float netHeadYaw, float headPitch) {
   }

   @Override
   public void renderToBuffer(
      PoseStack poseStack, VertexConsumer vertexConsumer, int packedLight, int packedOverlay, float red, float green, float blue, float alpha
   ) {
      this.main.render(poseStack, vertexConsumer, packedLight, packedOverlay, red, green, blue, alpha);
   }
}
